### PR TITLE
[CodeGen][MISched] Handle empty sized resource usage.

### PR DESCRIPTION
**Pull request by michaelmaitland**
TargetSchedule.td explicitly allows the usage of a ProcResource for zero cycles, in order to represent that the ProcResource must be available but is not consumed by the instruction. On the other hand, ResourceSegments explicitly does not allow for a zero sized interval. In order to remedy this, this patch handles the special case of when there is an empty interval usage of a resource by not adding an empty interval.

No test was added since there is no upstream CPU that does this currently. We ran into this issue downstream, but it makes sense to have this upstream since it is explicitly allowed by TargetSchedule.td.

### CODE REVIEW

**Comment by michaelmaitland** [comment]
ping

**Review by preames — APPROVED** [review]
LGTM

**Review by fpetrogalli — COMMENTED** [review]
Thank you for your patch, I agree that if the scheduler needs to know about empty resource usage, we need to take care of this.

In the commit message you say:
> In order to remedy this, this patch handles the special case of when there is a zero sized usage of a resource without creating a ResourceSegment for it.

Correct me if I am wrong, but this doesn't seem to be what your patch is doing?
A ResourceSegment is always created, even with your changes, it is just that you keep it empty (no intervals in it).

On a more general note. Would it be possible to have the scheduler deal with the situation "[...] that the ProcResource must be available but is not consumed by the instruction" by not even trying to add the interval? I am asking this because the situation in which we need to decide whether or not we can schedule an instruction in the case of "we have a resource that is not used by the instruction but needs to be available" seems to be different from the case  of "we have a resource that is used by the instruction", where we are forced to check whether or not we can schedule by checking the content of the ResourceSegment (empty/interval history overlapping the cycle in which the instruction will use that resource). 

Francesco

**Comment by fpetrogalli** [comment]
I actually have another question, which challenges this design:

> TargetSchedule.td explicitly allows the usage of a ProcResource for zero cycles, in order to represent that the ProcResource must be available but is not consumed by the instruction. 

Why is this allowed? And where is this specified? I could not find any reference of this in TargetSchedule.td

It seems to me that, if a resource is required to be available, it means that the instruction needs to use it? Why not just mark it for  at least 1 cycle usage?

Francesco

**Comment by michaelmaitland** [comment]
> A ResourceSegment is always created, even with your changes, it is just that you keep it empty (no intervals in it).

You are right that a ResourceSegment is always created, even with my changes, but it has no interval in it. I will reword the commit message accordingly.

**Comment by fpetrogalli** [comment]
> > It seems to me that, if a resource is required to be available, it means that the instruction needs to use it? Why not just mark it for at least 1 cycle usage?
> 
> Hi @fpetrogalli I have just run spec2006 comparing scheduler that uses `ReleaseAtCycle = [0]` scheduler that uses `sed ReleaseAtCycle = [0]/ReleaseAtCycle = [1]` on one of our downstream scheduler model. Unfortunately I am seeing significant regressions on spec2006 with the change. I am also skeptical about making this chance since it will force all other scheduler models that use this to retune their model.
> 

I am not surprised by this result. You are marking that cycle as using the resource, therefore there are less empty spaces that can be filled. 

It is clear that specifying a resource that have zero usage has a meaning in the scheduler, and this information is used to produce better schedules. I do not know where this logic is in the scheduler, I just find it wrong that such logic is taken all the way down to the resource tracking algorithm, because there is nothing (no interval) to track.

> As a result, I am in favor of this patch instead of a patch that imposes a `ReleaseAtCycle > 0` requirement. WDYT?

Yeah, restructuring the way the "resource used for 0 cycles" is used in the scheduler seems to be a piece of major rework, which is likely not to be worth the effort. At the end of the day, scheduling is not an exact science :)

However, please;

1. add a comment explaining that adding a 0-cycle resource does not make much sense in the resource segment, and the it is there just to fulfil some extra logic of the scheduler
2. please take care of the extra `|| CutOff == 0` condition, as I think it is not necessary (I might be wrong)

**Comment by michaelmaitland** [comment]
> I just find it wrong that such logic is taken all the way down to the resource tracking algorithm, because there is nothing (no interval) to track.

I agree with you, and this may be something we can discourage in the future, until we finally have a good enough justification to spend time to remove it entirely.

> However, please; ...

Updated :)

File: llvm/unittests/CodeGen/SchedBoundary.cpp [hunk 1]
```diff
@@ -20,9 +20,10 @@ TEST(ResourceSegmentsDeath, FullOverwrite) {
   EXPECT_DEATH(X.add({15, 18}), "A resource is being overwritten");
 }
 
-TEST(ResourceSegmentsDeath, ZeroSizeIntervalsNotAllowed) {
+TEST(ResourceSegmentsDeath, ZeroSizeCuttoffAllowed) {
   auto X = ResourceSegments({{10, 20}});
-  EXPECT_DEATH(X.add({20, 30}, 0), "0-size interval history has no use.");
+  X.add({20, 30}, 0);
```
Comment (fpetrogalli) [inline llvm/unittests/CodeGen/SchedBoundary.cpp:25]: With the latest version, I would expect this to fail on debug builds.

File: llvm/lib/CodeGen/MachineScheduler.cpp [hunk 2]
```diff
@@ -4266,6 +4266,12 @@ unsigned ResourceSegments::getFirstAvailableAt(
   assert(std::is_sorted(std::begin(_Intervals), std::end(_Intervals),
                         sortIntervals) &&
          "Cannot execute on an un-sorted set of intervals.");
+
+  // Zero resource usage is allowed by TargetSchedule.td but we do not construct
+  // a ResourceSegment interval for that situation.
+  if (AcquireAtCycle == Cycle)
+    return Cycle;
```
Comment (fpetrogalli) [inline llvm/lib/CodeGen/MachineScheduler.cpp:4273]: This seems to be wrong. Say that `CurrCycle` is 200, and that ResourceX have a `[AcquireAtCycle,ReleaseAtCycle]` pair `[5,5]` (note that `Cycle == ReleaseAtCycle` in the naming of the method's parameters... unfortunate)

Your code will say that the resource is available at cycle 5 (not even considering if the resource is busy at cycle 200), but in reality we can only schedule from cycle 200 onwards.

Comment (michaelmaitland) [reply llvm/lib/CodeGen/MachineScheduler.cpp]: I have updated to return CurrCycle, which would be 200 in this case. It should return CurrCycle since it is as if the for loop below hit the continue statement every time since nothing intersects with the empty interval.

Comment (michaelmaitland) [reply llvm/lib/CodeGen/MachineScheduler.cpp]: I will update Cycle -> ReleaseAtCycle in a future patch.

File: llvm/lib/CodeGen/MachineScheduler.cpp [hunk 3]
```diff
@@ -4266,6 +4266,12 @@ unsigned ResourceSegments::getFirstAvailableAt(
   assert(std::is_sorted(std::begin(_Intervals), std::end(_Intervals),
                         sortIntervals) &&
          "Cannot execute on an un-sorted set of intervals.");
+
+  // Zero resource usage is allowed by TargetSchedule.td but we do not construct
+  // a ResourceSegment interval for that situation.
```
Comment (fpetrogalli) [inline llvm/lib/CodeGen/MachineScheduler.cpp:4271]: > It should return CurrCycle since it is as if the for loop below hit the continue statement every time since nothing intersects with the empty interval.

I see, so this is an optimisation of the algorithm, not a functional change. It would be best not to add such optimisations (more code, more bugs!), unless there is a significant slow down in compile time.

If you want this optimisation, I would rephrase the comment as 

```suggestion
  // Empty resource intervals (`AcquireAtCycle == ReleaseAtCycle`)
  // are not stored in a ResourcesSegment.
  // However, zero resource usage is allowed by TargetSchedule.td.
  // If we deal with an empty interval, we can just return CurrCycle since
  // it is as if the for loop below hit the continue statement every time, 
  // since nothing intersects with the empty interval.
```

**Review by fpetrogalli — APPROVED** [review]
One more comment, either remove the "empty interval optimisation", or expand the comment to make clear it is an optimization and not a functional change.

Approved, thank you for your patience!

File: llvm/lib/CodeGen/MachineScheduler.cpp (hunk 3 above)
Comment (michaelmaitland) [reply llvm/lib/CodeGen/MachineScheduler.cpp:4271]: This is not an optimization. It was made because intersect will crash when considering the empty interval since intervals are supposed to be [a, b), not (a, b)

**Comment by michaelmaitland** [comment]
> One more comment, either remove the "empty interval optimisation", or expand the comment to make clear it is an optimization and not a functional change.

Hi @fpetrogalli, I made a comment in the thread above explaining why it is more than just an optimization.

**Comment by nickdesaulniers** [comment]
We're seeing build failures as a result of this change, PTAL: https://lab.llvm.org/buildbot/#/builders/225/builds/30132

**Comment by michaelmaitland** [comment]
> We're seeing build failures as a result of this change, PTAL: https://lab.llvm.org/buildbot/#/builders/225/builds/30132

I've pushed a fix in d2d42dcfde7e9bf58e49b376ae7488ffac97f4c8

File: llvm/lib/CodeGen/MachineScheduler.cpp [hunk 4]
```diff
@@ -4266,6 +4266,12 @@ unsigned ResourceSegments::getFirstAvailableAt(
...
+  if (AcquireAtCycle == Cycle)
```
Comment (nickdesaulniers) [inline llvm/lib/CodeGen/MachineScheduler.cpp:4272]:
```
llvm-project/llvm/lib/CodeGen/MachineScheduler.cpp:4281:25: error: use of undeclared identifier 'Cycle'
  if (AcquireAtCycle == Cycle)
                        ^
```

https://lab.llvm.org/buildbot/#/builders/225/builds/30132

Comment (nickdesaulniers) [reply llvm/lib/CodeGen/MachineScheduler.cpp:4272]: weird, isn't `Cycle` the function parameter? Did a rebase get messed up when landing?

Comment (michaelmaitland) [reply llvm/lib/CodeGen/MachineScheduler.cpp:4272]: I committed a rename from Cycle -> ReleaseAtCycle which included the name of the arg in the function. I forgot to rebase this patch on top of that and I introduced a use of `Cycle` in this patch.
